model-mesh-ovms-adapter/server/modelmanager.go

```diff
@@ -0,0 +1,499 @@
+		basePath:    basePath,
+		c:           c,
+		ctx:         ctx,
+	}
+
+	select {
+	case err := <-c:
+		if err != nil {
+			return fmt.Errorf("LoadModel failed: %w", err)
+		}
+		return nil
+	case <-ctx.Done():
+		return fmt.Errorf("LoadModel request was cancelled")
+	}
+}
+
+func (mm *OvmsModelManager) UnloadModel(ctx context.Context, modelId string) error {
+	c := make(chan error, 1)
+	mm.requests <- &request{
+		requestType: unload,
+		modelId:     modelId,
+		c:           c,
+		ctx:         ctx,
+	}
+
+	select {
+	case err := <-c:
+		if err != nil {
+			return fmt.Errorf("UnloadModel failed: %w", err)
+		}
+		return nil
+	case <-ctx.Done():
+		return fmt.Errorf("UnloadModel request was cancelled")
+	}
+}
+
+func (mm *OvmsModelManager) UnloadAll(ctx context.Context) error {
+	c := make(chan error, 1)
+	mm.requests <- &request{
+		requestType: unloadAll,
+		c:           c,
+		ctx:         ctx,
+	}
+
+	select {
+	case err := <-c:
+		if err != nil {
+			return fmt.Errorf("UnloadAll failed: %w", err)
+		}
+		return nil
+	case <-ctx.Done():
+		return fmt.Errorf("UnloadAll request was cancelled")
+	}
+}
+
+func (mm *OvmsModelManager) GetConfig(ctx context.Context) error {
+	return mm.getConfig(ctx)
+}
+
+// internal
+
+type requestType int64
+
+const (
+	load requestType = iota
+	unload
+	unloadAll
+)
+
+type request struct {
+	requestType requestType
+
+	modelId  string // for load and unload
+	basePath string // for load
+
+	ctx context.Context
+	c   chan<- error
+}
+
+// Run loop for the manager's internal actor that owns the model repository config
+//
+//
+// Maintains a slice of batched requests that are in process in the reload
+//
+// Returns results from the reload operation once it completes
+// Receives a stream of requests from its channel
+func (mm *OvmsModelManager) run() {
+	log := mm.log.WithValues("thread", "run")
+	log.Info("Starting ModelManger thread")
+	for mm.requests != nil {
+		// read a batch of requests to process
+		// we need a criteria to determine when to stop batching
+		// requests; here we take the approach of waiting for a fixed
+		// period of time after a request is received
+		batch := make([]*request, 0, 10)
+		var stopChan <-chan time.Time = nil
+
+	runLoopSelect:
+		select {
+		case req, ok := <-mm.requests:
+			if !ok {
+				// shutdown the loop after processing this batch
+				mm.requests = nil
+				break // from select
+			}
+			// add request to batch
+			batch = append(batch, req)
+
+			// set the stop timer, if not set already
+			if stopChan == nil {
+				stopChan = time.NewTimer(100 * time.Millisecond).C
+			}
+
+			goto runLoopSelect
+		case <-stopChan:
+			break // from select
+		}
+
+		log.V(1).Info("Processing batch of requests", "numRequests", len(batch))
+
+		// update the model configuration based on the requests
+		var unloadAllRequest *request
+
+		// build a new map to track changes to the same model
+		modelUpdates := map[string]*request{}
+		for _, req := range batch {
```

Comment:
> I was thinking it makes sense to just update loadedModelsMap as the requests come in. Then, if you haven't yet recorded any updates in the current batch for a particular model id, and the incoming change matches the state already in loadedModelsMap, it's effectively a no-op and can be completed immediately and otherwise ignored (e.g. wouldn't cause the timer to start if it wasn't already).

This assumes that the state recorded in loadedModelsMap is already synced with OVMS, which I don't know is safe in error cases (eg. if OVMS crashes and is in the process of reloading the models). Presumably, ModelMesh wouldn't send many spurious calls to load/unload a model that is already loaded/unloaded. This is potentially an abnormal state, and I think it'd be best to complete the reload. When handling something that may be abnormal (interleaved requests, load for loaded model, unload for unloaded model, etc.), I think it would be better not to assume that the state the adapter has cached is correct.

Reply:
> This assumes that the state recorded in loadedModelsMap is already synced with OVMS, which I don't know is safe in error cases 

This is kind of linked to the other [comment](https://github.com/kserve/modelmesh-runtime-adapter/pull/18#discussion_r836938486) above. I agree it should represent the target state but where we determine/return that models failed to load after the reload operation, I'm suggesting that we prune them since from our pov the target state no longer includes them. So I think that it _is_ safe and we'll always keep the config file on disk in sync with it so if OVMS container restarts it will load the correct version.

> If there was any change to the model spec (eg. path) the model id would be different (so the paths on disk would be different), right? I'm not seeing why extra precautions would be necessary.

In theory yes but that might not necessarily be the case for "general" (non kserve) model-mesh use. Where the same model id could be used with a different model path as long as the old one was first deleted. It's also a matter of being defensive from the adapter's side.